Allow user to set preference if they want tree list to be expanded by default

ericp-mrel opened the pull request:
This change implements the ability for a user in the back-end to be able to set a preference where they can decide if they want a tree list to be expanded by default or not. This change also makes it much easier to be able to open/close all tree list nodes (assuming they don't have a preference set by opening/closing them manually).

The initial default of the tree list is being expanded or not is inherited from the list config.

![Screenshot_2020-06-01 Category list OctoberCMS(2)](https://user-images.githubusercontent.com/63664924/83458196-abba1400-a427-11ea-9a1d-6a003ac28ab3.png)

Tree nodes are expanded once changing the setting.
![Screenshot_2020-06-01 Category list OctoberCMS(1)](https://user-images.githubusercontent.com/63664924/83458159-92b16300-a427-11ea-914e-fa6915b6fb4d.png)

Commit:
Allow user to set preference if they want tree list to be expanded by default.

Commit:
Allow user to set preference if they want tree list to be expanded by default.

LukeTowers commented:
@ericp-mrel I like the concept, but could we put it in the list header instead as a expand / collapse all button next to the check box?

Commit:
Add ability to forget backend widget session data.

Commit:
Add ability to forget backend widget session data.

Commit:
Add tree expand/collapse toggle button in list header.

Commit:
Add tree expand/collapse toggle button in list header.

Commit:
Revert "Allow user to set preference if they want tree list to be expanded by default."

This reverts commit a5eab782

Commit:
Revert "Allow user to set preference if they want tree list to be expanded by default."

This reverts commit a5eab782

Commit:
align tree list expand/collapse button in header with first level.

Commit:
align tree list expand/collapse button in header with first level.

LukeTowers commented:
Just post a screenshot when you're ready.

ericp-mrel commented:
Sorry this is taking so long, I've got a lot going on and haven't been able to find time to finish implementing the rest of this.

Once I add the last part of functionality I'll mark this is ready for merging. The last part I need to implement is detecting if all tree nodes have been collapsed and then making sure the toggle all icon switches to expand all nodes so the user doesn't need to click the button twice to collapse all first and then expand all.

Commit:
Make the toggle button always collapse tree nodes unless all nodes are collapsed.
When all tree nodes are collapsed, the toggle button will then expand all tree nodes.

Commit:
Make the toggle button always collapse tree nodes unless all nodes are collapsed.
When all tree nodes are collapsed, the toggle button will then expand all tree nodes.

Commit:
Add some comments.

Commit:
Add some comments.

ericp-mrel commented:
@LukeTowers  I think I've finally got this is a state I am satisfied with.

When default state of `$treeExpanded = false`, clicking the toggle tree button will expand all nodes and set the `$treeExpanded = true` preference.

![image](https://user-images.githubusercontent.com/63664924/86298149-b00e5400-bbc2-11ea-9840-5a56bce3d433.png)

Then, when at least one node is expanded, the toggle tree button will change to collapse all tree nodes.

![image](https://user-images.githubusercontent.com/63664924/86298219-dc29d500-bbc2-11ea-9e8b-d336af2dc3c6.png)

If all nodes are manually collapsed, the `$treeExpanded = false` preference will be set and the toggle tree button will then change to expanding all nodes.

![image](https://user-images.githubusercontent.com/63664924/86298382-3c207b80-bbc3-11ea-9bd5-56f22d9a5008.png)


I decided to make the toggle tree button default to collapsing nodes when possible since that seems to be the most common UX pattern when working with tree lists when I was looking around.

Let me know what you think and if there's anything else you think should be adjusted/changed.

Commit:
move `left` property into existing selector.

Commit:
move `left` property into existing selector.

mjauvin commented on modules/backend/traits/SessionMaker.php:
```diff
@@ -57,6 +58,21 @@ protected function getSession($key = null, $default = null)
+        Arr::forget($currentStore, $keys);
+
+        Session::put($sessionId, base64_encode(serialize($currentStore)));
+    }
```
```suggestion
    protected function forgetSession($keys)
    {
        $sessionId = $this->makeSessionId();
        $currentStore = $this->getSession();
        Arr::forget($currentStore, $keys);
        Session::put($sessionId, base64_encode(serialize($currentStore)));
    }
```

mjauvin commented on modules/backend/traits/SessionMaker.php:
```diff
@@ -57,6 +58,21 @@ protected function getSession($key = null, $default = null)
+    /**
+     * Removes one or many items from the widget session data.
+     * @param string|array $keys Unique key for the data store.
+     */
```
```suggestion
     * @return void
     */
```

mjauvin commented on modules/backend/widgets/Lists.php:
```diff
@@ -8,6 +8,7 @@
 use Carbon\Carbon;
 use October\Rain\Html\Helper as HtmlHelper;
 use October\Rain\Router\Helper as RouterHelper;
+use October\Rain\Support\Str;
```
```suggestion
use Str;
```

mjauvin commented on modules/backend/widgets/Lists.php:
```diff
@@ -1723,9 +1736,98 @@ public function isTreeNodeExpanded($node)
+            $this->toggleTreeExpanded = true;
+            $this->putSession('tree_toggle_expanded', $this->toggleTreeExpanded);
+        }
+
```
The previous if/elseif code block should be restructured to NOT repeat the code twice.

ericp-mrel replied:
What do you think about writing like this? I'm not sure this block is common enough to really needing it to be in its own method.

```suggestion
    public function onToggleTreeNode()
    {
        $this->putSession('tree_node_status_' . post('node_id'), post('status') ? 0 : 1);

        $expandNode = !post('status');

        if ((!$this->treeExpanded && $expandNode) || (!$expandNode && $this->isAllNodesCollapsed())) {
            $this->treeExpanded = false;
            $this->putSession('tree_expanded', $this->treeExpanded);
        }

        if (!$this->treeExpanded && $expandNode) {
            /*
             * When the tree is collapsed and the user expands a node,
             * make sure the tree expanded preference is set to false
             * and the toggle button is set to collapse all nodes.
             * The toggle nodes button should always default to collapsing nodes,
             * unless all nodes are already collapsed.
             */

            $this->toggleTreeExpanded = false;
            $this->putSession('tree_toggle_expanded', $this->toggleTreeExpanded);
        } elseif (!$expandNode && $this->isAllNodesCollapsed()) {
            /*
             * If the user is collapsing a node and all nodes are now collapsed,
             * make sure the tree expanded preference is set,
             * and change the toggle button to expand all nodes.
             */

            $this->toggleTreeExpanded = true;
            $this->putSession('tree_toggle_expanded', $this->toggleTreeExpanded);
        }

        return $this->onRefresh();
    }
```

mjauvin replied:
In both versions, the logic seems wrong to me. There has to be a simpler / cleaner / more readable way to write this, no?

mjauvin replied:
But if I have to choose one version, I prefer the former.

mjauvin replied:
Yeah, looking at it again, the previous version is fine, sorry.

mjauvin commented on modules/backend/widgets/Lists.php:
```diff
@@ -1723,9 +1736,98 @@ public function isTreeNodeExpanded($node)
+
+            $this->toggleTreeExpanded = false;
+            $this->putSession('tree_toggle_expanded', $this->toggleTreeExpanded);
+        } elseif (! $expandNode && $this->isAllNodesCollapsed()) {
```
```suggestion
        } elseif (!$expandNode && $this->isAllNodesCollapsed()) {
```

mjauvin commented on modules/backend/widgets/Lists.php:
```diff
@@ -1723,9 +1736,98 @@ public function isTreeNodeExpanded($node)
+
+        $expandNode = ! post('status');
+
+        if (! $this->treeExpanded && $expandNode) {
```
```suggestion
        if (!$this->treeExpanded && $expandNode) {
```

mjauvin commented on modules/backend/widgets/Lists.php:
```diff
@@ -1692,6 +1701,10 @@ public function validateTree()
 
+        $this->treeExpanded = $this->getSession('tree_expanded', $this->treeExpanded);
+
+        $this->toggleTreeExpanded = $this->getSession('tree_toggle_expanded', ! $this->treeExpanded);
```
```suggestion
        $this->toggleTreeExpanded = $this->getSession('tree_toggle_expanded', !$this->treeExpanded);
```

mjauvin commented on modules/backend/widgets/Lists.php:
```diff
@@ -1723,9 +1736,98 @@ public function isTreeNodeExpanded($node)
+        $this->treeExpanded = $this->toggleTreeExpanded;
+        $this->putSession('tree_expanded', $this->toggleTreeExpanded);
+
+        $this->toggleTreeExpanded = ! $this->toggleTreeExpanded;
```
```suggestion
        $this->toggleTreeExpanded = !$this->toggleTreeExpanded;
```

ericp-mrel committed:
Update modules/backend/traits/SessionMaker.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/traits/SessionMaker.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/traits/SessionMaker.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/traits/SessionMaker.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

ericp-mrel committed:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

mjauvin commented on modules/backend/widgets/Lists.php:
```diff
@@ -1723,9 +1736,98 @@ public function isTreeNodeExpanded($node)
+     */
+    protected function resetAllTreeNodes()
+    {
+        if (! $this->showTree) {
```
```suggestion
        if (!$this->showTree) {
```

mjauvin commented on modules/backend/widgets/Lists.php:
```diff
@@ -1723,9 +1736,98 @@ public function isTreeNodeExpanded($node)
     {
         $this->putSession('tree_node_status_' . post('node_id'), post('status') ? 0 : 1);
+
+        $expandNode = ! post('status');
```
```suggestion
        $expandNode = !post('status');
```

Commit:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

Commit:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

Commit:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

Commit:
Update modules/backend/widgets/Lists.php

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

LukeTowers commented on modules/backend/traits/SessionMaker.php:
```diff
@@ -1,5 +1,6 @@
 <?php namespace Backend\Traits;
 
+use October\Rain\Support\Arr;
```
```suggestion
use Arr;
```

LukeTowers commented on modules/backend/traits/SessionMaker.php:
```diff
@@ -57,6 +58,19 @@ protected function getSession($key = null, $default = null)
     }
 
+    /**
+     * Removes one or many items from the widget session data.
+     * @param string|array $keys Unique key for the data store.
+     * @return void
```
```suggestion
     * Removes one or many items from the widget session data.
     *
     * @param string|array $keys Keys to be removed from the widget session data
     * @return void
```

LukeTowers commented on modules/backend/widgets/Lists.php:
```diff
@@ -8,6 +8,7 @@
 use Carbon\Carbon;
 use October\Rain\Html\Helper as HtmlHelper;
 use October\Rain\Router\Helper as RouterHelper;
+use Str;
```
Move this above Html; please

LukeTowers commented on modules/backend/widgets/Lists.php:
```diff
@@ -1692,6 +1701,10 @@ public function validateTree()
 
         $this->showSorting = $this->showPagination = false;
 
+        $this->treeExpanded = $this->getSession('tree_expanded', $this->treeExpanded);
+
+        $this->toggleTreeExpanded = $this->getSession('tree_toggle_expanded', !$this->treeExpanded);
+
```
```suggestion
        $this->showSorting = $this->showPagination = false;
        $this->treeExpanded = $this->getSession('tree_expanded', $this->treeExpanded);
        $this->toggleTreeExpanded = $this->getSession('tree_toggle_expanded', !$this->treeExpanded);
```

Commit:
Update modules/backend/widgets/Lists.php

Commit:
Update modules/backend/widgets/Lists.php

ericp-mrel committed:
Apply suggestions from code review

Co-authored-by: Luke Towers <github@luketowers.ca>

ericp-mrel committed:
Apply suggestions from code review

Co-authored-by: Luke Towers <github@luketowers.ca>

Commit:
re-organize imports.

Commit:
re-organize imports.

LukeTowers commented:
@mjauvin what's the current status on this?

mjauvin commented:
@LukeTowers I don't know, I am not the author of the PR.

ericp-mrel commented:
This PR is done on my end since I've made all of the requested changes.
Unless anyone has any ideas on how to better simplify the logic for this, but this is about as simple as I could make it.

LukeTowers commented:
@ericp-mrel can you merge develop back into your branch? I'll review this in some more depth when I have some more time, probably after 1.1.2 is released.

Commit:
Merge remote-tracking branch 'origin/user-preference-tree-expanded' into user-preference-tree-expanded

ericp-mrel commented:
> @ericp-mrel can you merge develop back into your branch? I'll review this in some more depth when I have some more time, probably after 1.1.2 is released.

I've rebased onto the develop branch, but I had to do a merge before I was able to push to my feature branch. Hopefully I did that correctly.

LukeTowers commented:
@ericp-mrel looks like nope, it's screwed up. Please resubmit on a new branch fresh from the current develop branch.

daftspunk commented:
This doesn't seem like the right way to approach this. I've made a note to address the ability to expand/collapse all in our internal tracker. Thanks
